waffle-provider/src/MockProvider.ts

```diff
@@ -1,12 +1,36 @@
-import {providers, Wallet} from 'ethers';
+import {providers, Wallet, utils} from 'ethers';
 import {defaultAccounts} from './defaultAccounts';
 import Ganache from 'ganache-core';
 
 const defaults = {accounts: defaultAccounts};
 
+export interface RecordedCall {
+  readonly address: string;
+  readonly data: string;
+}
+
 export class MockProvider extends providers.Web3Provider {
+  private callHistory: RecordedCall[] = []
+
   constructor(private options?: Ganache.IProviderOptions) {
     super(Ganache.provider({...defaults, ...options}) as any);
+    this.recordCallHistory();
```

Comment:
You probably want a extract class `CallHistory`, so that we stick to SRP, and in the constructor you have:
```ts
this.callHistory = new CallHistory();
this.callHistory.record(this._web3Provider);
```